usm: http2: user mode dynamic table

guyarb opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->
### What does this PR do?

The PR introduces a new approach to manage our dynamic table.
The PR changes the minimum kernel version of HTTP2 monitoring to 5.4 (to support perf-event-output from socket filters)

The kernel sends to the user mode the conn-tuple, the index in the dynamic table and the path itself via a perf event. The streams no longer poses the request path, but just an index to the dynamic table. The user mode is responsible to resolve the indexes in to the actual paths.

In the kernel we have a "set" (hash map, the key is conn-tuple + index, and the value is meaningless, we care only if the entry exists in the map), that tells the kernel code if a given conn-tuple and index represents an indexed path or not.

The user mode LRU evicts the least recently used element when it is full, and removes the appropriate entry from the kernel set. The user mode LRU is a little bit smaller (by 10 entries) from the kernel map, to ensure the kernel map is never full.

The LRU maps conn-tuple and index to a pointer of string interner, to reduce redundant duplication.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

Before this PR we used a kernel maps to manage http2 dynamic table, at first we used LRU hash map,
but we found out the kernel LRU is not stable, and does not guarantee to perform a single eviction when the map is full, but can evict many items even if the map is not near to be full.
Then we used a simple hash map with a user-mode cleaner (when a connection is terminated), but we saw the maps are being full and connections live for a long period which means we lose our ability to insert new values to the map.

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

Kernel map changes:

`http2_dynamic_table` removed - saving 2.1875 MB (constant 10K entries)
`http2_interesting_dynamic_table_set` added - cost 4MB (for 65K entries by default, 2MB for production size)
`http2_in_flight` reduced from 15MB to 5.5MB for 65K

total saving 7.7MB

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

guyarb committed:
usm: http2: changed minimum kernel version for http2 processing to 5.4.0

guyarb committed:
usm: http2: Add user mode consumer to the perf buffer

The dynamic table management is now consuming the paths sent from the kernel.
We use an LRU datastore to automatically evict the least recently used element
when the map is full. On eviction we're deleting the matching entry from
http2_interesting_dynamic_table_set map.

The LRU value is a string internning pointer, to ensure we're using
a single instance and memory for the same path, even if the key is
different.

guyarb committed:
usm: http2: introduce new incomlete buffer

The new implementation keeps transactions as is (and not divide to requests / responses)
as the kernel implemenation sends only full transacations as a single unit.

guyarb committed:
usm: http2: Changed dynamic table map cleaner to clean the new set

As our goal is to remove the dynamic table map, we're changing the user mode map cleaner to
clean the interesting dynamic table set

guyarb committed:
usm: http2: Clean datastore when removing terminated connections

Adding to the map cleaner a post-cleanup operation, to remove items from the datastore
if their connection was terminated

guyarb committed:
usm: http2: tests: Modify test to verify the new feature works as expected

amitslavin commented:
Adding here a comment for us to remember adding load test results and dogfooding

guyarb commented on pkg/network/protocols/http2/model_linux.go:
```diff
@@ -182,7 +182,7 @@ func (tx *ebpfTXWrapper) String() string {
 	output.WriteString(fmt.Sprintf("[%s] [%s ⇄ %s] ", tx.family(), tx.sourceEndpoint(), tx.destEndpoint()))
 	output.WriteString(" Method: '" + tx.Method().String() + "', ")
-	buf := make([]byte, len(tx.Stream.Request_path))
+	buf := make([]byte, 2*maxHTTP2Path)
```
addressed in https://github.com/DataDog/datadog-agent/pull/21701

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding.h:
```diff
@@ -276,6 +276,7 @@ static __always_inline __u8 filter_relevant_headers(struct __sk_buff *skb, skb_i
 // process_headers processes the headers that were filtered in filter_relevant_headers,
 // looking for requests path, status code, and method.
 static __always_inline void process_headers(struct __sk_buff *skb, dynamic_table_index_t *dynamic_index, http2_stream_t *current_stream, http2_header_t *headers_to_process, __u8 interesting_headers,  http2_telemetry_t *http2_tel) {
+    bool dummy_value = true;
```
the value is not important (it can be false as well), we only for the existence in the map

guyarb commented on pkg/network/protocols/http2/dynamic_table.go:
```diff
@@ -111,8 +152,94 @@ func (dt *DynamicTable) setupDynamicTableMapCleaner(mgr *manager.Manager, cfg *c
+// addPath adds a new path to the dynamic table and the string interner.
+func (dt *DynamicTable) addPath(key http2DynamicTableIndex, path []byte) {
+	val := dt.stringInternStore.GetString(string(path))
+	dummyValue := true
```
the value is not important (it can be false as well), we only for the existence in the map

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding-defs.h:
```diff
@@ -56,13 +56,9 @@
-#define HTTP_INDEX_PATH     "\x60\xd5\x48\x5f\x2b\xce\x9a\x68"
-#define HTTP_INDEX_PATH_LEN (sizeof(HTTP_INDEX_PATH) - 1)
+// Special indexes to indicate the root path and the index path.
+#define HTTP_ROOT_PATH_INDEX    0xffffffffffffff00
```
done

guyarb committed:
usm: http2: Add path index

Adding to the current logic the path index, which will allow us to resolve the path (in
a future commit) in the user mode, rather than managing it in the kernel.

For the ROOT and INDEX paths, we're introducing constant (high) values, to mark them,
as both exists in the static table with a very low index, in which can cause us to mix
between new dynamic values with low index, and those paths.

guyarb committed:
usm: http2: Add interesting dynamic table hash map

The map acts as a set, to indicate if a given dynamic index (conn tuple + index) is interesting.
If a key exists - the index is interesting, otherwise it is not.

In a future commit we'll remove the old http2_dynamic_table map, and keep the new set instead
for kernel synchronization

guyarb committed:
usm: http2: Use user mode LRU to store and resolve paths

The user mode is no longer processing paths in processHTTP2, but relies on
our LRU map and its perf-buffer to do so. Now we're only resolving a path
by its connection tuple and index

guyarb committed:
usm: http2: Remove http2_dynamic_table

No longer is use

guyarb committed:
usm: http2: Remove redundant heap map

guyarb committed:
usm: http2: Add cyclic map

guyarb committed:
usm: http2: Send paths from kernel to user mode

guyarb committed:
usm: http2: remove redundant instance of dynamic_table_index_t in http2_ctx_t

guyarb committed:
usm: http2: Remove http2_ctx as it is not needed anymore

After the last 2 commits, the struct http2_ctx_t was just a wrapper to http2_stream_key_t.

guyarb committed:
wip: merge with above x2

guyarb committed:
usm: http2: Remove request path and path size from http2_stream_t

guyarb committed:
wip: tests

guyarb committed:
wip: dynamic table

guyarb committed:
wip: dynamic
